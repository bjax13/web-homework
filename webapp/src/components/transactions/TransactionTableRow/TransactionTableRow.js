import React from 'react'
import { any, arrayOf, string, bool, number, shape, func } from 'prop-types'
// import { css } from '@emotion/core'
import TransactionTableRowItem from '../TransactionTableRowItem/TransactionTableRowItem'
import { makeDataTestId, convertToRoman } from '../TransactionTable.utils'
import { AMOUNT_TYPE, DEBIT_TYPE, CREDIT_TYPE } from './TransactionTableRow.utils'
import Button from '../../general/Button/Button'

// const styles = css` //TODO time permitting add CSS
//   .header {
//     font-weight: bold;
//   }
// `

function TransactionTableRow ({ rowCells, transaction, onEdit, onDelete, isRomanNumeral }) {
  const { id: transactionId } = transaction

  // TODO time permitting add logic to combine Debit and Credit Columns
  const getTableValue = (rowItem) => {
    switch (rowItem.cell.type) {
      case AMOUNT_TYPE:
        return isRomanNumeral ? convertToRoman(rowItem.cell?.value) : rowItem.cell?.value
      case DEBIT_TYPE:
        return rowItem.cell?.value ? 'Debit' : ''
      case CREDIT_TYPE:
        return rowItem.cell?.value ? 'Credit' : ''
      default:
        return rowItem.cell?.value
    }
  }

  return (
    <>
      <tr data-testid={`transaction-${transactionId}`} key={`transaction-${transactionId}`}>
        {rowCells.map((rowItem) => (
          <TransactionTableRowItem dataTestid={rowItem.dataTestid} key={rowItem.dataTestid}>
            {getTableValue(rowItem)}
          </TransactionTableRowItem>
        ))}
        <TransactionTableRowItem dataTestid={makeDataTestId(transactionId, 'onEdit')}>
          <Button onClick={() => onEdit(transaction)}>Edit</Button>
        </TransactionTableRowItem>
        <TransactionTableRowItem dataTestid={makeDataTestId(transactionId, 'onDelete')}>
          <Button onClick={() => onDelete(transactionId)}>Delete</Button>
        </TransactionTableRowItem>
      </tr>
    </>
  )
}

TransactionTableRow.propTypes = {
  rowCells: arrayOf(
    shape({
      dataTestid: string,
      cell: shape({
        type: string,
        value: any
      })
    })
  ),
  transaction: shape({
    id: string,
    user_id: string,
    description: string,
    merchant_id: string,
    debit: bool,
    credit: bool,
    amount: number
  }),
  onEdit: func,
  isRomanNumeral: bool,
  onDelete: func
}

export default TransactionTableRow
