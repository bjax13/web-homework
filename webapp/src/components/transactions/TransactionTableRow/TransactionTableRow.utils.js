export const AMOUNT_TYPE = 'amount'
export const DEBIT_TYPE = 'debit'
export const CREDIT_TYPE = 'credit'
