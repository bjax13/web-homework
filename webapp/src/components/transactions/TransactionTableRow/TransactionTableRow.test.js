import '@testing-library/jest-dom'
import React from 'react'
import { IntlProvider } from 'react-intl'
import TransactionTableRow from './TransactionTableRow'
import { render, fireEvent, getByText } from '@testing-library/react'

const mockCells = [
  {
    dataTestid: 'transaction-61add94066be4c252121c140-id',
    cell: {
      type: 'id',
      value: '61add94066be4c252121c140'
    }
  },
  {
    dataTestid: 'transaction-61add94066be4c252121c140-description',
    cell: {
      type: 'description',
      value: 'Cleora buying Bacon'
    }
  },
  {
    dataTestid: 'transaction-61add94066be4c252121c140-debit',
    cell: {
      type: 'debit',
      value: false
    }
  },
  {
    dataTestid: 'transaction-61add94066be4c252121c140-credit',
    cell: {
      type: 'credit',
      value: true
    }
  },
  {
    dataTestid: 'transaction-61add94066be4c252121c140-amount',
    cell: {
      type: 'amount',
      value: 235
    }
  }
]

const mockTransaction = {
  __typename: 'Transaction',
  id: '61add94066be4c252121c140',
  user_id: null,
  description: 'Cleora buying withdrawal',
  merchant_id: null,
  debit: false,
  credit: true,
  amount: 235
}

describe('Transactions Table', () => {
  const mockOnEdit = jest.fn()
  const mockOnDelete = jest.fn()

  const sharedProps = {
    isRomanNumeral: false,
    onDelete: mockOnDelete,
    onEdit: mockOnEdit,
    rowCells: mockCells,
    transaction: mockTransaction
  }

  test('renders table', async () => {
    const { container } = render(
      <IntlProvider locale='en'>
        <TransactionTableRow {...sharedProps} />
      </IntlProvider>
    )

    expect(container).toMatchSnapshot()
  })

  test('renders table row with buttons', async () => {
    const { container } = render(
      <IntlProvider locale='en'>
        <TransactionTableRow {...sharedProps} />
      </IntlProvider>
    )

    fireEvent.click(getByText(container, 'Edit'))
    expect(mockOnEdit).toHaveBeenCalled()
    fireEvent.click(getByText(container, 'Delete'))
    expect(mockOnDelete).toHaveBeenCalled()
  })
})
