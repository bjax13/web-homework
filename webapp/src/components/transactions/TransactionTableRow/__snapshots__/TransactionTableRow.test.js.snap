// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Transactions Table renders table 1`] = `
.emotion-0 {
  border: 0;
  background: #78f89f;
  border-radius: 5px;
  padding: 0.5rem 1rem;
  margin: 0.2rem;
  font-size: 0.8rem;
  line-height: 1;
}

<div>
  <tr
    data-testid="transaction-61add94066be4c252121c140"
  >
    <td
      data-testid="transaction-61add94066be4c252121c140-id"
    >
      61add94066be4c252121c140
    </td>
    <td
      data-testid="transaction-61add94066be4c252121c140-description"
    >
      Cleora buying Bacon
    </td>
    <td
      data-testid="transaction-61add94066be4c252121c140-debit"
    />
    <td
      data-testid="transaction-61add94066be4c252121c140-credit"
    >
      Credit
    </td>
    <td
      data-testid="transaction-61add94066be4c252121c140-amount"
    >
      235
    </td>
    <td
      data-testid="transaction-61add94066be4c252121c140-onEdit"
    >
      <button
        class="emotion-0 emotion-1"
      >
        Edit
      </button>
    </td>
    <td
      data-testid="transaction-61add94066be4c252121c140-onDelete"
    >
      <button
        class="emotion-0 emotion-1"
      >
        Delete
      </button>
    </td>
  </tr>
</div>
`;
